File: android/src/main/kotlin/com/cleveradssolutions/plugin/flutter/CASBridgeBuilder.java
package com.cleveradssolutions.plugin.flutter;

import android.app.Activity;

import androidx.annotation.NonNull;

import com.cleveradssolutions.plugin.flutter.bridge.MediationManagerMethodHandler;
import com.cleversolutions.ads.ConsentFlow;
import com.cleversolutions.ads.android.CAS;

import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNull;

import kotlin.jvm.functions.Function0;

public final class CASBridgeBuilder {
    final CAS.ManagerBuilder builder;
    final @NotNull Function0<Activity> activityProvider;
    final @NotNull CASInitCallback initCallback;

    public CASBridgeBuilder(
            @NotNull Function0<Activity> activityProvider,
            @NotNull CASInitCallback initCallback
    ) {
        this.activityProvider = activityProvider;
        builder = CAS.buildManager();
        this.initCallback = initCallback;
    }

    public void withTestMode(boolean isEnabled) {
        builder.withTestAdMode(isEnabled);
    }

    public void setUserId(String id) {
        builder.withUserID(id);
    }

    public void disableConsentFlow() {
        builder.withConsentFlow(new ConsentFlow(false));
    }

    public void enableConsentFlow(String url) {
        builder.withConsentFlow(new ConsentFlow().withPrivacyPolicy(url));
    }

    public void addExtras(String key, String value) {
        builder.withMediationExtras(key, value);
    }

    @NonNull
    @Contract("_, _, _, _, _ -> new")
    public CASBridge build(
            @NonNull String id,
            @NonNull String flutterVersion,
            int formats,
            ConsentFlow flow,
            MediationManagerMethodHandler mediationManagerMethodHandler
    ) {
        return buildInternal(id, flutterVersion, formats, flow, mediationManagerMethodHandler);
    }

    @NonNull
    @Contract("_, _, _, _, _ -> new")
    public CASBridge buildInternal(
            @NonNull String id,
            @NonNull String flutterVersion,
            int formats,
            ConsentFlow flow,
            MediationManagerMethodHandler mediationManagerMethodHandler
    ) {
        builder.withEnabledAdTypes(formats)
                .withCasId(id)
                .withFramework("Flutter", flutterVersion)
                .withConsentFlow(flow);
        return new CASBridge(activityProvider, this, mediationManagerMethodHandler);
    }
}

File: ios/Classes/CASBridgeBuilder.swift
//
//  CASBridgeBuilder.swift
//  clever_ads_solutions
//
//  Copyright © 2024 CleverAdsSolutions LTD, CAS.AI. All rights reserved.
//

import CleverAdsSolutions
import Foundation

public class CASBridgeBuilder /* : FlutterCaller*/ {
//    var flutterCaller: completion?

    let rootViewController: UIViewController

    let managerBuilder: CASManagerBuilder

    let initCallback: CASInitCallback

    init(_ rootViewController: UIViewController, _ initCallback: CASInitCallback) {
        self.rootViewController = rootViewController

//        consent = CASConsentFlow()
//            .withViewControllerToPresent(rootViewController)
        managerBuilder = CAS.buildManager() // .withConsentFlow(consent)

        self.initCallback = initCallback
    }

//    func setFlutterCaller(caller: @escaping(completion)) {
//        flutterCaller = caller
//        self.consentHandler = { status in
//            var args = [String: Any]()
//            args["status"] = status.rawValue
//            self.flutterCaller?("OnDismissListener", args)
//        }
//        consent.withCompletionHandler(consentHandler)
//    }

    func withTestMode(isEnabled: Bool) {
        managerBuilder.withTestAdMode(isEnabled)
    }

    func setUserId(id: String) {
        managerBuilder.withUserID(id)
    }

    func addExtras(keyString: String, valueString: String) {
        managerBuilder.withMediationExtras(valueString, forKey: keyString)
    }

    func build(
        id: String,
        flutterVersion: String,
        formats: CASTypeFlags,
        consentFlow: CASConsentFlow,
        mediationManagerMethodHandler: MediationManagerMethodHandler
    ) -> CASBridge {
        return buildInternal(id: id, flutterVersion: flutterVersion, formats: formats, consentFlow: consentFlow, mediationManagerMethodHandler: mediationManagerMethodHandler)
    }

    func buildInternal(
        id: String,
        flutterVersion: String,
        formats: CASTypeFlags,
        consentFlow: CASConsentFlow,
        mediationManagerMethodHandler: MediationManagerMethodHandler
    ) -> CASBridge {
        managerBuilder.withAdFlags(formats)
        managerBuilder.withFramework("Flutter", version: flutterVersion)

        let bridge = CASBridge(builder: self, casID: id, mediationManagerMethodHandler: mediationManagerMethodHandler)
        bridge.getManager().adLoadDelegate = bridge
        return bridge
    }
}
